### PR TITLE
Integrated tests, gas profiling and fees fixes

### LINKED ISSUE #82: [MVP] End-to-end QA — recurring Bank messages
This ticket involves writing a basic guide so that we can test recurring Bank messages.

So:

1. production build command
2. Add alice, bob, and charlie users if they don't exist
3. deploy command using `junod` or `wasmd`
4. instantiate
5. Add task from alice that pays bob and charlie .000001 JUNO every hour (or 5 minutes, whatever) (can use https://crontab.guru)
6. set up agent using https://github.com/CronCats/croncat-rs
7. watch agent call proxy_call
8. Confirm it repeats
9. Confirm the balance decreases on the task
10. Confirm bob and charlie get paid

### LINKED ISSUE #83: [MVP] End-to-end testnet QA — simplest Task with 1 rule
This ticket involves writing a basic guide so that we can test that the rules are working.

So:

1. production build command
2. Add alice if they don't exist
3. deploy command using `junod` or `wasmd`
4. instantiate
5. Add task from alice that has 1 rule, checking whether a given DAO DAO proposal's status is "passed" (or whatever that Enum value is. (See https://github.com/DA0-DA0/dao-contracts/blob/cef33570e41c52b04acd692cdd527c836185fd13/packages/voting/src/status.rs#L7-L21)
6. set up agent using https://github.com/CronCats/croncat-rs
7. watch agent call proxy_call, giving the task hash if it evaluates to true
8. Confirm the transaction is successful
9. Confirm the DAO DAO proposal is executed.

We can use this testnet DAO DAO DAO:
https://testnet.daodao.zone/dao/juno1p2fuar474uv2p6vfnr2eu4nv9rx4m0qd26xuygkesje0r6pzhrssvrnh2y

### LINKED ISSUE #117: [Important] Address agent fee being high
As we discussed, the agent fee is too high so we'll want to make some changes to this.

I think it'll be easiest to test this all out on a local chain using the new scripts Raf and Nino have made.

Misha pointed out this line may want to be adjusted:
https://github.com/CronCats/cw-croncat/blob/e0a4e5fed2cc7bf827752114acd8bade6a809c3d/contracts/cw-croncat/src/contract.rs#L60

We can also consider adjusting the defaults set during instantiation, like:
https://github.com/CronCats/cw-croncat/blob/e0a4e5fed2cc7bf827752114acd8bade6a809c3d/contracts/cw-croncat/src/contract.rs#L19

This might be another thing to look at:
https://github.com/CronCats/cw-croncat/blob/e0a4e5fed2cc7bf827752114acd8bade6a809c3d/contracts/cw-croncat/src/contract.rs#L61

And, more unlikely, we can take a look at:
https://github.com/CronCats/cw-croncat/blob/e0a4e5fed2cc7bf827752114acd8bade6a809c3d/contracts/cw-croncat/src/contract.rs#L59

### LINKED ISSUE #125: Integrated tests and gas profiling
I'll use [cosm_orc](https://github.com/de-husk/cosm-orc) for this task, so we can setup [CI](https://github.com/de-husk/cosm-orc-gas-diff-action) later.
It helps to fix #117, #82, #83

### PR DESCRIPTION
**Pull request by Buckram123**
Closes #125, fixes #117, resolves #83, resolves #82

### CODE REVIEW

**Comment by TrevorJTClarke** [comment]
Progress is lookin solid!

**Comment by Buckram123** [comment]
Fixing #137 and it's ready for review

File: ci/gas-benchmark/src/test_cases.rs [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        interval: Interval::Immediate,
+        boundary: None,
+        stop_on_fail: false,
+        actions: vec![delegate_to_validator],
+        rules: None,
+        cw20_coins: vec![],
+    }
+}
+
+pub(crate) fn delegate_to_validator_twice(denom: &str) -> TaskRequest {
```
Comment (TrevorJTClarke) [inline ci/gas-benchmark/src/test_cases.rs:250]: Is the goal this should fail? double delegation seems odd

File: contracts/cw-croncat/src/agent.rs [hunk 2]
```diff
@@ -146,10 +146,11 @@ impl<'a> CwCroncat<'a> {
         // Check if native token balance is sufficient for a few txns, in this case 4 txns
         // TODO: Adjust gas & costs based on real usage cost
         let agent_wallet_balances = deps.querier.query_all_balances(account.clone())?;
-        let unit_cost = c.gas_price * 4;
+        let gas_cost = calculate_required_amount(c.gas_action_fee, c.agent_fee)?;
+        let unit_cost = c.gas_fraction.calculate(4 * gas_cost, 1)?;
```
Comment (TrevorJTClarke) [inline contracts/cw-croncat/src/agent.rs:150]: where does `4` come from? Maybe i forgot something?

File: contracts/cw-croncat/src/contract.rs [hunk 3]
```diff
@@ -8,15 +8,21 @@ use cosmwasm_std::{
...
 const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
 const DEFAULT_NOMINATION_DURATION: u16 = 360;
 
-// default for juno
-pub(crate) const GAS_BASE_FEE_JUNO: u64 = 400_000;
+/// default for juno
+/// This based on non-wasm operations, wasm ops seem impossible to predict
+pub const GAS_BASE_FEE_JUNO: u64 = 300_000;
+/// Gas cost per single action
+pub const GAS_ACTION_FEE_JUNO: u64 = 200_000;
```
Comment (TrevorJTClarke) [inline contracts/cw-croncat/src/contract.rs:22]: how did you arrive at this value?

File: justfile [hunk 4]
```diff
@@ -0,0 +1,36 @@
+test_addrs := env_var_or_default('TEST_ADDR', `jq -r '.[].address' ci/test_accounts.json | tr '\n' ' '`)
```
Comment (TrevorJTClarke) [inline justfile:1]: what is this file? Was it meant to get committed?

**Review by TrevorJTClarke — APPROVED** [review]
This is lookin awesome!
Love the cleanup and refactors im seeing.
Left a few questions, so im understanding whats changing & how on the exact fee pricing.
Only thing missing is another set of tests for Rules tasks, but that can be a separate issue.

File: ci/gas-benchmark/src/test_cases.rs (hunk 1 above)
Comment (Buckram123) [reply ci/gas-benchmark/src/test_cases.rs:250]: > Is the goal this should fail? double delegation seems odd

Just delegating twice, so we can get approximate gas usage per "delegate action"

File: contracts/cw-croncat/src/agent.rs (hunk 2 above)
Comment (Buckram123) [reply contracts/cw-croncat/src/agent.rs:150]: Found the origin of this `4`:
https://github.com/CronCats/cw-croncat/commit/4353f397c07f4480c5c679ea63a19de669317dad#diff-3bab5cd53a1040a4a531afe1501403786256e9dea8731dc20028c9beabddd2d3R111

I guess the idea here we wanted agent to have a balance enough to complete around 4 `proxy_call`'s at the time of registration. Agree about `4` being is way too "magic" number,  do we need this check? We can add new optional field to the config and take this value from there, thoughts? @TrevorJTClarke

File: justfile (hunk 4 above)
Comment (deveusss) [reply justfile:1]: @Buckram123

Comment (Buckram123) [reply justfile:1]: Yeah this is very cool [tool](https://github.com/casey/just), it's like a `Makefile` but imho better. Noticed that we have some boilerplate inside scripts, so the end goal is to move the most common commands there. 

It's already in use in the `CI` for running `gas-benchmark`

#### And the most important for the user:
running `just optimize`(or other commands we might end up finding useful) looks much easier to me than searching where to copy from that docker command, thoughts?

**Review by mikedotexe — APPROVED** [review]

File: justfile (hunk 4 above)
Comment (mikedotexe) [reply justfile:1]: Let's MEEEERGE and continue the discussion in tomorrow's meeting :D

**Comment by mikedotexe** [comment]
Insanely good work

**Pull request merged**